viz_scripts/scaffolding.py

```diff
@@ -105,46 +105,144 @@ def data_quality_check(expanded_ct):
 
 def unit_conversions(df):
     df['distance_miles']= df["distance"]*0.00062 #meters to miles
+    df['duration_h'] = df['duration'] / 60 / 60 #seconds to hours
+
+
+def feat_eng(data, const, feats, prefs, mode='Mode_confirm', repm='Replaced_mode'):
+    """
+    Calculate trip aggregate results from constants and append to data (Feature Engine)
+
+    Parameters:
+        data - trip data from OpenPATH
+        const - Pandas DataFrame with constant values for each mode (requires 'mode' feature)
+        feats - python list of feature names in const DataFrame that are of interest
+        prefs - prefixes to append to current feature names for new feature names
+        mode - feature name in data of feature with confirmed mode
+        repm - feature name in data of feature with replaced mode 
+
+    Returns:
+        data with appended features for each trip for both mode and replaced mode
+    """
+
+    # Check that const has a mode feature
+    if('mode' not in const.columns):
+        print('Error: mode not in constants dataframe.')
+        return data
+
+    # Check features list and prefix list same length
+    if(len(feats) != len(prefs)):
+        print("Error: prefix list and feature list not the same length.")
+        return data
+
+    # Check all feature names in constants dataframe
+    for feat in feats:
+        if(feat not in const.columns):
+            print('Error: ' + feat + ' not in constants dataframe.')
+            return data
+
+    # Use copies, don't change originals
+    data = data.copy()
+    const = const.copy()
+
+    # Duplicate mode feature in constant dataframe
+    for m in [mode, repm]:
+        const[m] = const['mode']
+
+    # Feature engine!
+    for i in range(len(feats)):
+        for m in [mode, repm]:
+            dic = dict(zip(const[m],const[feats[i]]))
+
+            # Create new feature in data
+            fn = prefs[i]+m
+            data[fn] = data[m].map(dic)
+            print('Created ' + fn + ' feature in data.')
```

Comment:
This is a bit tricky, so it would be nice to add some examples here - e.g.

```suggestion
            # Create new feature in data
            # prefs are [`cost__trip_`] or similar, and the mode and replaced_mode strings are passed in from the call 
            # location, so fn is `cost__trip_Mode_confirm`
            fn = prefs[i]+m
            # maps data["Mode_confirm"] against the dictionary
            # and then sets `cost__trip_Mode_confirm` to it.
            data[fn] = data[m].map(dic)
            print('Created ' + fn + ' feature in data.')
```